### DISCUSSION TITLE
`map` with multiple input matrices / arrays

### DISCUSSION THREAD

**Discussion by dvd101x** [josdejong/mathjs#3196] [Ideas]
Hi,

Currently `map` uses a single input matrix or array and a callback function with a single input

```js
const square = math.square
const x = [1, 2, 3]
math.map(x, square) // [1, 4, 9]
```

What if it could work with multiple inputs?
```js
const sumOfSquares = (a, b) => a**2 + b**2
const x = [1, 2, 3]
const y = [4, 5, 6]
math.map(x, y, sumOfSquares) // [17, 29, 45]
```
And include broadcasting
```js
const sumOfSquares = (a, b) => a**2 + b**2
const x = [1, 2]
const y = [[3], [4]]
math.map(x, y, sumOfSquares) // [[10, 13], [17, 20]]
```
I would like to discuss the idea of `map` using multiple broadcasted inputs.

**Comment by josdejong** [comment]
Nice idea, I like it 👍

**Reply by dvd101x** [reply]
I'm glad you like it.

Is it ok if I attempt a PR for this new functionality?

**Reply by josdejong** [reply]
Yes, that sounds good!

**Reply by dvd101x** [reply]
Hi Jos, 

I started to look into the arguments of the callback function.

> the callback is invoked with three arguments: the value of the entry, the index at which that entry occurs, and the full matrix/array being traversed

But with N matrices or arrays being mapped, it would be like so.

| N | callback |
|---|---|
| 1 | `callback(value, index, array)` |
| 2 | `callback(valueA, valueB, index, broadcastedArrayA, broadcastedArrayB)` |
| 3 | `callback(valueA, valueB, valueC, index, broadcastedArrayA, broadcastedArrayB, broadcastedArrayC)`|

With this change it would be something like

> the callback is invoked with 2N+1 arguments: the N values of the entry, the index at which that entry occurs, and the N full broadcasted matrix/array being traversed where N is the number of matrices being mapped.

I would like to validate in the meantime if this form makes the most sense.

**Reply by josdejong** [reply]
Your proposed API makes sense indeed. It looks a bit unwieldy, but in practice, the `broadcastedArrayA, broadcastedArrayB, broadcastedArrayC` part is not often needed, and this very order makes most sense since it is consistent with the case of 1 matrix. In typical usage, it will be `callback(valueA, valueB, valueC)` and sometimes `callback(valueA, valueB, valueC, index)`, which is easy to follow.

**Comment by lulunac27a** [comment]
It's based on [`Array.prototype.map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) function in JavaScript

**Reply by dvd101x** [reply]
Thank you for the reference.

**Comment by josdejong** [comment]
This is implemented now in `v13.1.0`, see #3228.
